pkg/controller/che/oauth_initial_htpasswd_provider.go

```diff
@@ -0,0 +1,245 @@
+//
+
+package che
+
+import (
+	"context"
+
+	errorMsg "errors"
+
+	"github.com/eclipse/che-operator/pkg/deploy"
+	"github.com/eclipse/che-operator/pkg/util"
+	configv1 "github.com/openshift/api/config/v1"
+	oauthv1 "github.com/openshift/api/config/v1"
+	userv1 "github.com/openshift/api/user/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	htpasswdIdentityProviderName        = "htpasswd-eclipse-che"
+	htpasswdSecretName                  = "htpasswd-eclipse-che"
+	ocConfigNamespace                   = "openshift-config"
+	openShiftOAuthUserCredentialsSecret = "openshift-oauth-user-credentials"
+)
+
+// OpenShiftOAuthUserHandler - handler to create or delete new Openshift oAuth user.
+type OpenShiftOAuthUserHandler interface {
+	CreateOAuthInitialUser(userName string, crNamespace string, openshiftOAuth *oauthv1.OAuth) error
+	DeleteOAuthInitialUser(userName string, crNamespace string) error
+}
+
+// OpenShiftOAuthUserOperatorHandler - implementation OpenShiftOAuthUserHandler.
+type OpenShiftOAuthUserOperatorHandler struct {
+	OpenShiftOAuthUserHandler
+	runtimeClient client.Client
+	runnable      util.Runnable
+}
+
+// NewOpenShiftOAuthUserHandler - create new OpenShiftOAuthUserHandler instance
+func NewOpenShiftOAuthUserHandler(runtimeClient client.Client) OpenShiftOAuthUserHandler {
+	return &OpenShiftOAuthUserOperatorHandler{
+		runtimeClient: runtimeClient,
+		// real process implementation. In the test we are using mock
+		runnable: util.NewRunnable(),
+	}
+}
+
+// CreateOAuthInitialUser - creates new htpasswd provider with inital user with Che flavor name
+// if Openshift cluster hasn't got identity providers, otherwise do nothing.
+// It usefull for good first user expirience.
+// User can't use kube:admin or system:admin user in the Openshift oAuth. That's why we provide
+// initial user for good first meeting with Eclipse Che.
+func (iuh *OpenShiftOAuthUserOperatorHandler) CreateOAuthInitialUser(userName string, crNamespace string, openshiftOAuth *oauthv1.OAuth) error {
+	password := util.GeneratePasswd(6)
+	htpasswdFileContent, err := iuh.generateHtPasswdUserInfo(userName, password)
+	if err != nil {
+		return err
+	}
+
+	secret := &corev1.Secret{}
+	nsName := types.NamespacedName{Name: htpasswdSecretName, Namespace: ocConfigNamespace}
+	if err := iuh.runtimeClient.Get(context.TODO(), nsName, secret); err != nil {
+		if errors.IsNotFound(err) {
+			htpasswdFileSecretData := map[string][]byte{"htpasswd": []byte(htpasswdFileContent)}
+			if err = deploy.CreateSecret(htpasswdFileSecretData, htpasswdSecretName, ocConfigNamespace, iuh.runtimeClient); err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	if err := appendIdentityProvider(openshiftOAuth, iuh.runtimeClient); err != nil {
```

Comment:
If identityProvider exists ?

Reply:
I applied check to prevent applying identity provider few times.